futures-util/src/future/future/shared.rs

```diff
@@ -156,6 +158,35 @@ where
     pub fn weak_count(&self) -> Option<usize> {
         self.inner.as_ref().map(Arc::weak_count)
     }
+
+    /// Hashes the internal state of this `Shared` in a way that's compatible with `ptr_eq`.
+    pub fn ptr_hash<H: Hasher>(&self, state: &mut H) {
+        match self.inner.as_ref() {
+            Some(arc) => {
+                state.write_u8(1);
+                ptr::hash(Arc::as_ptr(arc), state);
+            }
+            None => {
+                state.write_u8(0);
+            }
+        }
+    }
+
+    /// Returns `true` if the two `Shared`s point to the same future (in a vein similar to
+    /// `Arc::ptr_eq`).
+    ///
+    /// Returns `false` if either `Shared` has terminated.
```

Comment:
The implementation that returns false if both are terminated would appear incompatible with ptr_hash above. See also docs: https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq

That said, it also seems odd that it always returns true when both are terminated.

Reply:
I don't understand this comment. This `ptr_eq` returns *false* if one or both `Shared`s are terminated. And the constraint on `Hash` is that if k1 == k2 then hash(k1) == hash(k2), it says nothing about k1 != k2.

Reply:
Ok, I think it was my misunderstanding.